Support -NoProfile via checkbox

csharmath opened the pull request:
[Updating the PR description to reflect all the changes]

This change will add an option for PowerShell via a checkbox to use the -NoProfile switch when executing.
The rationale behind this change is to allow PowerShell executing without any profile in a clean shell.

Currently this is not possible so many of us needing -NoProfile have to use windows batch to start powershell with -NoProfile ( or pwsh from bash).

Additionally, based on comments on this PR these have also been changed
* support -NoProfile for both Windows PowerShell.exe and Linux pwsh
* changed to -File instead of the implied -Command "& script.ps1" version as it returns the proper exit code
* removed quoting of the -File parameter as it caused Jenkins failure on Windows - this might need more follow-up
* Since the plugin now supports both Windows and Linux I've renamed "Windows PowerShell" to just "PowerShell" to reflect that it's now cross platform => Does it have a potential of breaking something ? as if yes, I would undo that part.

I've built and tested the plugin locally on both Windows and Linux agents and confirmed that it worked.

Maintainers: I would recommend taking the offer from @rkeithhill who is a well known [PowerShell OSS](https://github.com/Pscx/Pscx) [contributor](https://stackoverflow.com/users/153982/keith-hill) and [book author](https://rkeithhill.wordpress.com/2009/03/08/effective-windows-powershell-the-free-ebook/)


>Not that I need another OSS project to work on, I currently work on posh-git, Plaster, vscode-powershell and PowerShellEditorServices but ... do you guys need help maintaining this repo? We rely on this plugin heavily where I work so I'm motivated to keep it updated and of high quality. But I'd need a little help on what is required to build, deploy and test it locally.

Keith: this is how you test it:
* spin up locally 
` java -jar jenkins.war --httpPort=5555`
* or just build the plugin
`mvn package`
* then upload the target/powershell.hpi to your test instance http://$jenkins:port/pluginManager/advanced

aheubusch commented:
Cloned the PR locally. I was able to build and run all tests successfully. I was also able to run jenkins and see the plugin in the plugin manager and in a freestyle project (with the checkbox checked).

csharmath commented:
Wow, that's awesome to hear! Thanks for checking it! 
I guess then my local dev setup is at fault, glad that I have asked!

csharmath commented:
I still have no luck building it unfortunately.

I keep getting the below error when trying to run "mvn package"
`Caused by: java.lang.AssertionError: class hudson.plugins.powershell.PowerShell is missing its descriptor`
...
`Tests in error:
  testConfigRoundtrip(hudson.plugins.powershell.PowerShellTest): 500 Internal Server Error for http://localhost:59904/job/test0/configSubmit`

do you mind sharing how you guys are building ?

My env on Windows 7

> java -version
java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)
> mvn -version
Apache Maven 3.5.0 (ff8f5e7444045639af65f6095c62210b5713f426; 2017-04-03T15:39:06-04:00)
Maven home: C:\dev\java\maven\bin\..
Java version: 1.8.0_131, vendor: Oracle Corporation
Java home: C:\Program Files\Java\jdk\jre
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 7", version: "6.1", arch: "amd64", family: "windows"

Also if I just "mvn package -o -DskipTests" and try to upload the plugin it doesn't show up on jenkins...

Thanks

aheubusch commented:
My stuff is a bit outdated, havent bothered to upgrade in a while
```
g:\Jenkins\powershell-plugin>java -version
java version "1.7.0_75"
Java(TM) SE Runtime Environment (build 1.7.0_75-b13)
Java HotSpot(TM) 64-Bit Server VM (build 24.75-b04, mixed mode)

g:\Jenkins\powershell-plugin>mvn -version
Apache Maven 3.3.3 (7994120775791599e205a5524ec3e0dfe41d4a06; 2015-04-22T13:57:37+02:00)
Maven home: g:\maven\apache-maven-3.3.3\bin\..
Java version: 1.7.0_75, vendor: Oracle Corporation
Java home: c:\Program Files\Java\jdk1.7.0_75\jre
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 8.1", version: "6.3", arch: "amd64", family: "windows"
```
Building is straight forward. `mvn test` to run the tests, and `mvn hpi:run` to start jenkins with the plugin auto installed (will also run the tests).

rkeithhill commented on src/main/webapp/help.html:
```diff
@@ -4,6 +4,8 @@
   The text you enter in the text box will be executed as a PowerShell file, and a build
   will be considered a failure if at the end of the execution $LastExitCode is not 0.
+
+  By default PowerShell uses profile(s) which can be disabled to avoid any potential conflict and provide a clean shell.
```
"By default, PowerShell processes profiles scripts at startup.  This can be disabled to improve startup time, avoid any potential conflict and provide a clean shell."

rkeithhill commented on src/main/java/hudson/plugins/powershell/PowerShell.java:
```diff
@@ -26,7 +35,11 @@ protected String getFileExtension() {
     public String[] buildCommandLine(FilePath script) {
         if (isRunningOnWindows(script)) {
-            return new String[] { "powershell.exe", "-NonInteractive", "-ExecutionPolicy", "Bypass", "& \'" + script.getRemote() + "\'"};
+            if (useProfile){
```
I was just thinking about adding this very feature.  Great to see you have already submitted a PR for it.  Any chance you could apply the same logic to the non-Windows case?  The same principle applies there.  If you rebased you'd pick up my fix to line 46 `powershell` -> `pwsh`.  Thanks!

rkeithhill commented:
@chrisalbrecht @olivergondza Any chance we could get this PR and #12 merged?  Actually this PR needs a little work, basically, just replicate the code change made in this PR to the `non-Windows` (else) case?  

Not that I need another OSS project to work on, I currently work on posh-git, Plaster, vscode-powershell and PowerShellEditorServices but ... do you guys need help maintaining this repo?  We rely on this plugin heavily where I work so I'm motivated to keep it updated and of high quality.  But I'd need a little help on what is required to build, deploy and test it locally.

csharmath commented:
@rkeithhill thanks for your feedback!
I've made my changes to support Linux and updated the description + synced from upstream.

rkeithhill commented on src/main/java/hudson/plugins/powershell/PowerShell.java:
```diff
@@ -43,7 +43,11 @@ protected String getFileExtension() {
+                return new String[] { "pwsh", "-NonInteractive", "& \'" + script.getRemote() + "\'"};
+            } else {
+                return new String[] { "pwsh", "-NonInteractive", "-NoProfile", "& \'" + script.getRemote() + "\'"};
+            }
```
👍

rkeithhill commented on src/main/webapp/help.html:
```diff
@@ -5,7 +5,8 @@
   will be considered a failure if at the end of the execution $LastExitCode is not 0.
 
-  By default PowerShell uses profile(s) which can be disabled to avoid any potential conflict and provide a clean shell.
+  By default, PowerShell processes profiles scripts at startup. 
```
Minor nit - `profiles` -> `profile`.  I think it will read better.  Thanks!

csharmath commented:
I've tested pwsh on Linux and didn't work as is with the below error:
```
Started by user anonymous
Building in workspace /u/harmath/git/powershell-plugin/work/jobs/test powershell/workspace
[workspace] $ pwsh -NonInteractive -NoProfile "& '/tmp/hudson5369535234780594662.ps1'"
The argument '& '/tmp/hudson5369535234780594662.ps1'' is not recognized as the name of a script file. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

Usage: pwsh[.exe] [[-File] <filePath> [args]]
                  [-Command { - | <script-block> [-args <arg-array>]
                                | <string> [<CommandParameters>] } ]
                  [-ConfigurationName <string>] [-CustomPipeName <string>]
                  [-EncodedCommand <Base64EncodedCommand>]
                  [-ExecutionPolicy <ExecutionPolicy>] [-InputFormat {Text | XML}]
                  [-Interactive] [-LoadProfile] [-NoExit] [-NoLogo] [-NonInteractive] [-NoProfile]
                  [-OutputFormat {Text | XML}] [-Version] [-WindowStyle <style>]
                  [-WorkingDirectory <directoryPath>]

       pwsh[.exe] -h | -Help | -? | /?

PowerShell Online Help https://aka.ms/pscore6-docs

All parameters are case-insensitive.
Build step 'Windows PowerShell' marked build as failure
Finished: FAILURE
```

Then I've checked and pwsh is indeed not working with single or double quote escaped file paths.
```
> pwsh --version
PowerShell 7.0.0-preview.1
> pwsh -NonInteractive -NoProfile -File ~/test.ps1
Test
> pwsh -NonInteractive -NoProfile -File '~/test.ps1'
The argument '~/test.ps1' is not recognized as the name of a script file. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

So I've changed the invocation of pwsh to 
* use -File
* and removed quoting the path

And now pwsh on Linux is working (not sure how did it actually work if at all before?)
![image](https://user-images.githubusercontent.com/1450460/61066105-f3304f80-a3d2-11e9-9dda-35c05c88679a.png)

While at it, I've also renamed "Windows PowerShell" to "PowerShell" as it's no longer just Windows only.
If that's causing an unintended side effect, please let me know and I will change it back.

csharmath commented:
I was also wondering if the script.getRemote() is doing proper cross platform Path handling ?
i.e. escaping spaces.

rkeithhill commented:
Yeah, switching to `-File` is the change that #12 made. This is better for pwsh (maybe even powershell) because you get back the actual exit code if someone used `host.setShouldExit(<number>)` instead of always `1` to indicate failure - which is how `-Command` works.

rkeithhill commented:
The problem with stripping the quoting is that paths with spaces will not work. I know that's not typically a problem on Linux but it is on Windows.
```
C:\temp
07-11 12:27:42 13ms 9> pwsh -file C:\temp\path with spaces\test.ps1
The argument 'C:\temp\path' is not recognized as the name of a script file. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```
But if we use quotes we give up globbing the `~`.  Another option is to escape PowerShell special chars including space i.e. this works:
```
pwsh -file C:\temp\path` with` spaces\test.ps1
Running in test.ps1
Saving command history...
```
But you have to escape at least space, [, ], $, @ and probably a few more.  FWIW -File only accepts a single file, I won't execute multiple files.

rkeithhill commented:
Then again, maybe this isn't really an issue because Jenkins controls the path used with `-File`.  On Linux it looks like it is always something like `/tmp/hudson<mumble>.ps1`.  I suspect it is a similar temp dir path.

BTW the reason this didn't work:
```
pwsh -NonInteractive -NoProfile "& '/tmp/hudson5369535234780594662.ps1'"
```
is because it needs to be:
```
pwsh -NonInteractive -NoProfile -Command "& '/tmp/hudson5369535234780594662.ps1'"
```

csharmath commented:
So what should be the best is if we use -File and single quote the path, but check this out:
```Java
if (isRunningOnWindows(script)) {
            if (useProfile){
                return new String[] { "powershell.exe", "-NonInteractive", "-ExecutionPolicy", "Bypass", "-File", "\'" + script.getRemote() + "\'"};
            } else {
                return new String[] { "powershell.exe", "-NoProfile", "-NonInteractive", "-ExecutionPolicy", "Bypass", "-File", "\'" + script.getRemote() + "\'"};
            }
        } else {
            // ExecutionPolicy option does not work (and is not required) for non-Windows platforms
            // See https://github.com/PowerShell/PowerShell/issues/2742
            if (useProfile){
                return new String[] { "pwsh", "-NonInteractive", "-File", "\'" + script.getRemote() + "\'"};
            } else {
                return new String[] { "pwsh", "-NonInteractive", "-NoProfile", "-File", "\'" + script.getRemote() + "\'"};
            }
        }
```

Will result in this on windows

```
Started by user unknown or anonymous
Running as SYSTEM
Building remotely on codebox (windows) in workspace c:\local\jenkins\workspace\test powershell
[test powershell] $ powershell.exe -NoProfile -NonInteractive -ExecutionPolicy Bypass -File 'C:\Users\harmath\AppData\Local\Temp\jenkins2917772564201870597.ps1'
Processing -File ''C:\Users\harmath\AppData\Local\Temp\jenkins2917772564201870597.ps1'' failed: The given path's format is not supported. Specify a valid path for the -File parameter.
Windows PowerShell 
Copyright (C) Microsoft Corporation. All rights reserved.

Build step 'PowerShell' marked build as failure
Finished: FAILURE
```

So it seems like that it's already quoted ?
I am rebuilding / retesting it now with 
```Java
return new String[] { "powershell.exe", "-NoProfile", "-NonInteractive", "-ExecutionPolicy", "Bypass", "-File", script.getRemote()};
```

Which might be the best bet here.
Especially that these path's are not user provided, but Jenkins internal path derived from temp env var.

csharmath commented:
>So it seems like that it's already quoted ?
I mean it's not double quoted on the web UI display, but it might be executed with double single quotes as otherwise it should have been working.

running some more tests and will probably check with procmon to see what's going on here.

csharmath commented:
There is no quoting with `"-File", script.getRemote()`
![image](https://user-images.githubusercontent.com/1450460/61078962-d9e9cc00-a3ef-11e9-80ec-f03679db6577.png)

And there is only a single quote with `"-File", "\'" + script.getRemote() + "\'"};`
![image](https://user-images.githubusercontent.com/1450460/61079158-495fbb80-a3f0-11e9-94c8-7057b171dee1.png)

But when executed by Jenkins it's not working
![image](https://user-images.githubusercontent.com/1450460/61079227-66948a00-a3f0-11e9-8bb3-5eedf4bdba8f.png)

If I executed the same thing in a shell it works
```
>'"hello"' | sc C:\Users\harmath\AppData\Local\Temp\jenkins16729219045772737193.ps1
>powershell.exe -NoProfile -NonInteractive -ExecutionPolicy Bypass -File 'C:\Users\harmath\AppData\Local\Temp\jenkins16729219045772737193.ps1'
hello
```

With all of this here is my take
* Jenkins controls the temporary ps1 file path via the tmp env var
* in my tests these env vars have no spaces on either Linux or Windows
* there is something funky with single quoted -File argument 
* not quoting just works 
* -File is better as stated by @rkeithhill above
* It seems to be safe to go with -File without quotes.

What others think? Do I miss something?
If we agree, I will update with the -File without quotes version.

rkeithhill commented:
Thanks for taking the time to do this. We could really use this update. Hopefully, the maintainers can get this merged in and published.

csharmath commented:
Okay, updated with `-File` and no quotes

csharmath commented:
Also updated the PR description based on all the chatter and changes

csharmath commented:
@rkeithhill - you can request to be added as maintainer here
https://wiki.jenkins.io/display/JENKINS/Adopt+a+Plugin#AdoptaPlugin-IknowwhichpluginIwanttohelpwith,whatshouldIdonow?

chrisalbrecht commented:
@rkeithhill can you rebase this and verify it's all still good to merge? I no longer have a java build environment set up to test this.

rkeithhill commented:
@csharmath Are you able to rebase your PR branch?
